core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java

```diff
@@ -158,11 +160,66 @@ RelSubset getOrCreateSubset(
       final VolcanoPlanner planner =
           (VolcanoPlanner) cluster.getPlanner();
 
+      // Converters from newly introduced subset to all the remaining one, only if possible to
+      // convert.  No point adding converters if it is not possible.
+      for (RelSubset other : subsets) {
+        if (other.getConvention() == Convention.NONE) {
+          continue;
+        }
+
+        assert other.getTraitSet().size() == subset.getTraitSet().size();
+
+        final ImmutableList<RelTrait> otherTrait =
+                other.getTraitSet().difference(subset.getTraitSet());
+        final ImmutableList<RelTrait> thisTrait =
+                subset.getTraitSet().difference(other.getTraitSet());
+        assert otherTrait.size() == thisTrait.size();
+
+        if (otherTrait.size() == 1 && thisTrait.size() == 1) {
+          RelTraitDef traitDef = otherTrait.get(0).getTraitDef();
+          boolean canConvert = traitDef.canConvert(
```

Comment:
I think that we should check satisfies here, rather than in canConvert. CanConvert should return true iff calling convert will return a result.